### ISSUE TITLE
 Locking behavior when setting properties conditionally with `MATCH` and `SET` leads to data being set unreliably on concurrent writes

### ISSUE THREAD

**Issue by maiieul** [neo4j/neo4j#12823]
I discovered that locking behavior when setting properties conditionally with `MATCH` and `SET` can lead to data being set unreliably on concurrent writes.

**Neo4j Version:** 4.4.0
**Operating System:** Linux Mint 20.1 Cinnamon
**API:** Neo4j Aura

### Steps to reproduce

1. Launch a blank neo4j sandbox; create following graph

```
CREATE (p:Post{title:"Hello", numberOfUpvotes: 0})
CREATE (u:User{firstname:"Maieul"})
MERGE (u)-[:UPVOTED_POST]->(p)
SET p.numberOfUpvotes = 1
RETURN p, u
```

2.  Run following query concurrently (you have 15 secs to run it at least twice :tongue:)

```
MATCH (p:Post{title:"Hello"})
MATCH (u:User{firstname:"Maieul"})
MATCH (u)-[up:UPVOTED_POST]->(p)
CALL apoc.util.sleep(15000)
DELETE up
SET p.numberOfUpvotes = p.numberOfUpvotes - 1
RETURN p
```

p.numberOfUpvotes will be decremented by the number of concurrent queries you ran.

### Expected behavior

p.numberOfUpvotes should be decremented only once, since subsequent queries shouldn't match the `(u)-[up:UPVOTED_POST]->(p)` pattern.

### Actual behavior

All concurrent queries will match the `(u)-[up:UPVOTED_POST]->(p)` pattern. 

This is what I believe happens: 

A first query among the concurrent queries will lock `(p)` to `SET p.numberOfUpvotes` and the others will wait. When the first query will be commited and unlock `(p)`, one of the waiting concurrent query will acquire the lock on `(p)` and proceed to `SET p.numberOfUpvotes` without checking first if `(u)-[up:UPVOTED_POST]->(p)` still matches. And its the same for following concurrent queries when they acquire the lock on `(p)`.

### Workaround

This is the best workaround I could find so far. 

```
MATCH (p:Post{title:"Hello"})
MATCH (u:User{firstname:"Maieul"})
CALL apoc.lock.nodes([p, u])
MATCH (u)-[up:UPVOTED_POST]->(p)
DELETE up
SET p.numberOfUpvotes = p.numberOfUpvotes - 1
RETURN p
```
I'm not sure if it's useful to lock `u`, but that's a detail I believe.

### Additional context

Above behavior seems to be especially problematic when a user would click on a button to upvote a post and click again to remove the upvote. Clicking quickly many times can lead to concurrent writes and thus to a wrong number of upvotes. 

It is possible to count the number of UPVOTED_POST relationships but having a numberOfUpvotes property helps with indexes and incrementing/decrementing it is more performant than counting the number of relationships on writes.

Lastly, I have another example case where I need to lock before a match without any DELETE clause. 

To reproduce:

1. create user, post and merge relationship
```
CREATE (p:Post{id:"1", numberOfUpvotes: 0})
CREATE (u:User{id:"2"})
MERGE (u)-[:UPVOTED_POST]->(p)
SET up.atDateTime = [datetime()], p.numberOfUpvotes = p.numberOfUpvotes + 1
RETURN p, u
```
2. Same as above, you can run it concurrently to see it increment or decrement more than once (here, the goal was to decrement p.numberOfUpvotes when the number of elements in the up.atDateTime list is even and increment p.numberOfUpvotes when it's odd).
```
MATCH (p:Post{id:"1"})
MATCH (u:User{id:"2"})
MATCH (u)-[up:UPVOTED_POST]->(p)
CALL apoc.util.sleep(15000)
SET 
    up.atDateTime = up.atDateTime + datetime(),
    p.numberOfUpvotes = CASE WHEN size(up.atDateTime)%2 = 0 THEN p.numberOfUpvotes - 1 ELSE p.numberOfUpvotes + 1 END
RETURN p
```
My workaround:

```
MATCH (p:Post{id:"1"})
MATCH (u:User{id:"2"})
CALL apoc.lock.nodes([p, u])
MATCH (u)-[up:UPVOTED_POST]->(p)
CALL apoc.util.sleep(15000)
SET 
    up.atDateTime = up.atDateTime + datetime(),
    p.numberOfUpvotes = CASE WHEN size(up.atDateTime)%2 = 0 THEN p.numberOfUpvotes - 1 ELSE p.numberOfUpvotes + 1 END
RETURN p
```

### Possible solution

Like MERGE, [performing another match afther locking on bound elements could solve the issue.](https://neo4j.com/developer/kb/understanding-how-merge-works/#_merge_acquires_locks_on_nodes_and_relationships_in_the_pattern_when_resulting_in_pattern_creation)

**Comment by HannesSandberg** [comment]
Thanks for the report. We'll take a look and get back to you.

**Comment by loveleif** [comment]
Hi @maieulChevalier, thanks again for reporting this issue and for doing so in such a clear manner! 

One thing to note is that transactions in Neo4j use a read-committed isolation level by default, which means they will see data as soon as it has been committed and will not see data in other transactions that have not yet been committed. As you have noted, one can manually acquire write locks on nodes and relationships to achieve higher level of isolation. You can find documentation about transaction management [here](https://neo4j.com/docs/java-reference/current/transaction-management/).

Except for your workaround, I believe it's also possible to guarantee the integrity of `numberOfUpvotes` by first doing a soft delete, from the same transaction:

```
// First, soft delete the relationship
MATCH (p:Post { title: 'Hello' })
MATCH (u:User { firstname: 'Maieul' })
MATCH (u)-[up:UPVOTED_POST]->(p)
SET up.deleted = true

// Then, in the same transaction(!), hard delete and update p.numberOfUpvotes
MATCH (p:Post { title: 'Hello' })
MATCH (u:User { firstname: 'Maieul' })
MATCH (u)-[up:UPVOTED_POST { deleted: true }]->(p)
DELETE up
SET p.numberOfUpvotes = p.numberOfUpvotes - 1
```

I tried this out, with added sleep, and one of the transactions fails and rollbacks. To try it out for yourself, you can open two instances of cypher-shell and run these concurrently:

```
:begin // Begin transaction

// First, soft delete the relationship
MATCH (p:Post { title: "Hello" })
MATCH (u:User { firstname: "Maieul" })
MATCH (u)-[up:UPVOTED_POST]->(p)
WHERE up.deleted IS NULL
SET up.deleted = true;

// Then, still in the same transaction, hard delete and update count
MATCH (p:Post {title:"Hello"})
MATCH (u:User {firstname:"Maieul"})
MATCH (u)-[up:UPVOTED_POST { deleted: true }]->(p)
CALL apoc.util.sleep(15000)
DELETE up
SET p.numberOfUpvotes = p.numberOfUpvotes - 1;

:commit // Commit transaction
```

**Comment by loveleif** [comment]
I'm closing the issue, as I believe the observed behaviour is correct. My previous comment also contains a solution to achieve what's wanted. Feel free re-open or create a new issue if anything is unclear.

**Comment by loveleif** [comment]
It might also be possible to re-calculate number of upvotes. Something like:

```
MATCH (p:Post { title: 'Hello' })
MATCH (u:User { firstname: 'Maieul' })
MATCH (u)-[up:UPVOTED_POST]->(p)
DELETE up
SET p.numberOfUpvotes = size(()-[:UPVOTED_POST]->(p))
```
